lib/mail_catcher/smtp.rb

```diff
@@ -1,58 +1,415 @@
+
+      attr :url, :options
+
+      def address
+        endpoint.address
+      end
+
+      def secure?
+        ['smtps'].include?(@url.scheme)
+      end
+
+      def protocol
+        Protocol::SMTP
+      end
+
+      def default_port
+        secure? ? 465 : 25
+      end
+
+      def default_port?
+        port == default_port
+      end
+
+      def port
+        @options[:port] || @url.port || default_port
+      end
+
+      def hostname
+        @options[:hostname] || @url.hostname
+      end
+
+      def scheme
+        @options[:scheme] || @url.scheme
+      end
+
+      def authority
+        if default_port?
+          hostname
+        else
+          "#{hostname}:#{port}"
+        end
+      end
+
+      def path
+        @url.path
+      end
+
+      LOCALHOST = 'localhost'
+
+      # We don't try to validate peer certificates when talking to localhost because they would always be self-signed.
+      def ssl_verify_mode
+        case hostname
+        when LOCALHOST
+          OpenSSL::SSL::VERIFY_NONE
+        else
+          OpenSSL::SSL::VERIFY_PEER
+        end
+      end
+
+      def ssl_context
+        @options[:ssl_context] || ::OpenSSL::SSL::SSLContext.new.tap do |context|
+          context.set_params(
+            verify_mode: ssl_verify_mode
+          )
+        end
+      end
+
+      def tcp_options
+        { reuse_port: @options[:reuse_port] ? true : false }
+      end
+
+      def build_endpoint(endpoint = nil)
+        endpoint ||= Async::IO::Endpoint.tcp(hostname, port, tcp_options)
+
+        if secure?
+          # Wrap it in SSL:
+          return Async::IO::SSLEndpoint.new(endpoint,
+                                            ssl_context: ssl_context,
+                                            hostname: hostname)
+        end
+
+        endpoint
+      end
+
+      def endpoint
+        @endpoint ||= build_endpoint
+      end
+
+      def bind(*args, &block)
+        endpoint.bind(*args, &block)
+      end
+
+      def connect(*args, &block)
+        endpoint.connect(*args, &block)
+      end
+
+      def each
+        return to_enum unless block_given?
+
+        endpoint.each do |endpoint|
+          yield self.class.new(@url, endpoint, @options)
+        end
+      end
+
+      def key
+        [@url.scheme, @url.userinfo, @url.host, @url.port, @options]
+      end
+
+      def eql?(other)
+        key.eql? other.key
+      end
+
+      def hash
+        key.hash
+      end
     end
 
-    super
-  end
+    Envelope = Struct.new(:sender, :recipients, :content)
 
-  def current_message
-    @current_message ||= {}
-  end
+    require 'async/io/protocol/line'
 
-  def receive_reset
-    @current_message = nil
-    true
-  end
+    module Protocol
+      module SMTP
+        def self.server(stream, *args)
+          Server.new(stream, *args)
+        end
 
-  def receive_sender(sender)
-    current_message[:sender] = sender
-    true
-  end
+        class Server < Async::IO::Protocol::Line
```

Comment:
Hi @sj26,
Can you suggest what type of spec we need for this? I am a little bit confused here 😅

Reply:
I'm not sure! How is async-http tested?

https://github.com/socketry/async-http

Reply:
The biggest challenge in switching smtp implementations is making sure we have good smtp client compatibility. I'm not yet sure how to achieve that.

Reply:
I have added RSpec for `smtp.rb` file. Do have a look.

Also, Does client compatibility means, the server should work with different clients available like in ruby we have `Net::SMTP`

Reply:
> The biggest challenge in switching smtp implementations is making sure we have good smtp client compatibility. I'm not yet sure how to achieve that.

I think we can release a beta version first & let the users test it, we can improve the implementation by the issue reports if we get any. Also, we can ask devs to use it by making topics on [meta](https://meta.discourse.org).